File: py/candlestick_patterns/patterns/evening_star.py
"""Evening Star pattern (3-candle bearish reversal)."""
from __future__ import annotations

from ..core.primitives import (
    is_white, is_black, real_body, is_real_body_gap_up,
)

EVENING_STAR_PENETRATION_FACTOR: float = 0.3


def evening_star(self) -> int:
    """Evening Star: a three-candle bearish reversal pattern.

    Must have:
    - first candle: long white real body,
    - second candle: short real body that gaps up (real body gap up from the
      first),
    - third candle: black real body that moves well within the first candle's
      real body.

    The meaning of "long" is specified with self._long_body.
    The meaning of "short" is specified with self._short_body.

    Returns:
        -100 for bearish, 0 for no pattern.
    """
    if not self._enough(3, self._long_body, self._short_body):
        return 0

    o1, h1, l1, c1 = self._bar(3)
    o2, h2, l2, c2 = self._bar(2)
    o3, h3, l3, c3 = self._bar(1)

    penetration = EVENING_STAR_PENETRATION_FACTOR

    if (is_white(o1, c1) and
            real_body(o1, c1) > self._avg(self._long_body, 3) and
            real_body(o2, c2) <= self._avg(self._short_body, 2) and
            is_real_body_gap_up(o1, c1, o2, c2) and
            is_black(o3, c3) and
            c3 < c1 - real_body(o1, c1) * penetration):
        return -100

    return 0

File: ts/candlestick-patterns/patterns/evening-star.ts
/** Evening Star pattern (3-candle bearish reversal). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, isWhite, isBlack, isRealBodyGapUp } from '../core/primitives.ts';

const EVENING_STAR_PENETRATION = 0.3;

/**
 * Evening Star: a three-candle bearish reversal pattern.
 *
 * Must have:
 * - first candle: long white real body,
 * - second candle: short real body that gaps up (real body gap up from the
 *   first),
 * - third candle: black real body that moves well within the first candle's
 *   real body.
 *
 * The meaning of "long" is specified with `longBody`.
 * The meaning of "short" is specified with `shortBody`.
 *
 * Category A: always bearish (continuous).
 *
 * Returns:
 *   Continuous float in [-100, 0].  Always bearish.
 */
export function eveningStar(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(3, cp.longBody, cp.shortBody)) return 0.0;

    const b1 = cp.bar(3);
    const b2 = cp.bar(2);
    const b3 = cp.bar(1);

    // Crisp gates: color checks and gap.
    if (!(isWhite(b1.o, b1.c) &&
        isRealBodyGapUp(b1.o, b1.c, b2.o, b2.c) &&
        isBlack(b3.o, b3.c))) return 0.0;

    // Fuzzy conditions.
    const muLong1 = cp.muGreaterCS(realBody(b1.o, b1.c), cp.longBody, 3);
    const muShort2 = cp.muLessCS(realBody(b2.o, b2.c), cp.shortBody, 2);

    // b3.c < b1.c - rb1 * penetration  →  b3.c < threshold
    const rb1 = realBody(b1.o, b1.c);
    const threshold = b1.c - rb1 * EVENING_STAR_PENETRATION;
    const width = cp.fuzzRatio * rb1 * EVENING_STAR_PENETRATION;
    const muPenetration = cp.muLtRaw(b3.c, threshold, width);

    return -tProductAll(muLong1, muShort2, muPenetration) * 100.0;
}
